src/System.Net.Mime/src/System/Net/Mime/ContentDisposition.cs

```diff
@@ -0,0 +1,342 @@
+            ParseValue();
+            headers.InternalSet(MailHeaderInfo.GetString(MailHeaderID.ContentDisposition), ToString());
+            _isPersisted = true;
+        }
+
+        internal void PersistIfNeeded(HeaderCollection headers, bool forcePersist)
+        {
+            if (IsChanged || !_isPersisted || forcePersist)
+            {
+                headers.InternalSet(MailHeaderInfo.GetString(MailHeaderID.ContentDisposition), ToString());
+                _isPersisted = true;
+            }
+        }
+
+        internal bool IsChanged => _isChanged || _parameters != null && _parameters.IsChanged;
+
+        public override string ToString()
+        {
+            if (_disposition == null || _isChanged || _parameters != null && _parameters.IsChanged)
+            {
+                _disposition = Encode(false); // Legacy wire-safe format
+                _isChanged = false;
+                _parameters.IsChanged = false;
+                _isPersisted = false;
+            }
+            return _disposition;
+        }
+
+        internal string Encode(bool allowUnicode)
+        {
+            var builder = new StringBuilder();
+            builder.Append(_dispositionType); // Must not have unicode, already validated
+
+            // Validate and encode unicode where required
+            foreach (string key in Parameters.Keys)
+            {
+                builder.Append("; ");
+                EncodeToBuffer(key, builder, allowUnicode);
+
+                builder.Append('=');
+                EncodeToBuffer(_parameters[key], builder, allowUnicode);
+            }
+
+            return builder.ToString();
+        }
+
+        private static void EncodeToBuffer(string value, StringBuilder builder, bool allowUnicode)
+        {
+            Encoding encoding = MimeBasePart.DecodeEncoding(value);
+            if (encoding != null) // Manually encoded elsewhere, pass through
+            {
+                builder.Append('"').Append(value).Append('"');
+            }
+            else if ((allowUnicode && !MailBnfHelper.HasCROrLF(value)) // Unicode without CL or LF's
+                || MimeBasePart.IsAscii(value, false)) // Ascii
+            {
+                MailBnfHelper.GetTokenOrQuotedString(value, builder, allowUnicode);
+            }
+            else
+            {
+                // MIME Encoding required
+                encoding = Encoding.GetEncoding(MimeBasePart.DefaultCharSet);
+                builder.Append('"').Append(MimeBasePart.EncodeHeaderValue(value, encoding, MimeBasePart.ShouldUseBase64Encoding(encoding))).Append('"');
+            }
+        }
+
+        public override bool Equals(object rparam)
+        {
+            return rparam == null ?
+                false :
+                string.Compare(ToString(), rparam.ToString(), StringComparison.OrdinalIgnoreCase) == 0;
+        }
+
+        public override int GetHashCode() => ToString().ToLowerInvariant().GetHashCode();
```

Comment:
Not identical. I see different values for the following (on desktop):

```c#
string foo = "Foo";
Console.WriteLine(foo.ToLowerInvariant().GetHashCode());
Console.WriteLine(StringComparer.OrdinalIgnoreCase.GetHashCode(foo));
Console.WriteLine(StringComparer.InvariantCultureIgnoreCase.GetHashCode(foo));
```

I didn't have easy access to `StringComparer.InvariantCultureIgnoreCase` in .NET Core, but using the equivalent `CultureInfo.InvariantCulture.CompareInfo.GetHashCode(foo, CompareOptions.IgnoreCase)` also gave differing results from `ToLowerInvariant` and `StringComparer.OrdinalIgnoreCase`.

You would think `StringComparer.InvariantCultureIgnoreCase` would give the same result as `ToLowerInvariant`, but it does not. I'm guessing it has something to do with `String.GetHashCode` using the randomized hash code, and a differing hash algorithm used in the comparers.

But given `OrdinalIgnoreCase` is being used for `Equals`, shouldn't it really be `OrdinalIgnoreCase` for `GetHashCode` to match?

Maybe we could consider the change separately.